### PR TITLE
GRU Sequential Surface Decoder: arc-length-ordered recurrent SRF

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current SRF (Surface Refinement Head) is an MLP that processes each surface node **independently** — it has no way to communicate context from the stagnation point to the suction peak node, or from the suction peak to the trailing edge recovery region. But airfoil pressure physics is **causally sequential along arc-length**: the stagnation point pressure sets the suction peak magnitude, which then determines the recovery gradient.

We hypothesize that replacing the SRF MLP with a **bidirectional GRU** that processes surface nodes in arc-length order will allow the decoder to propagate this causal structure, improving suction peak prediction accuracy (especially p_in and p_tan).

This is grounded in the Mamba Neural Operator (arXiv 2410.02113, JCP 2025) which shows SSM-based recurrence outperforms transformer self-attention for 2D PDE prediction. We use a GRU (simpler, no new packages) to test the same architectural hypothesis.

## Instructions

Replace the current SRF MLP head with a bidirectional GRU decoder. The GRU processes surface nodes in arc-length order (LE → upper surface → TE → lower surface → back to LE).

### Step 1: Sort surface nodes by arc-length

Before feeding surface tokens to the SRF, sort them by arc-length position around the airfoil. Arc-length order: start from leading edge (min X), go upper surface (positive Y), reach trailing edge (max X), go lower surface (negative Y), return to LE.

A simple proxy: sort by polar angle from centroid `θ = atan2(y - cy, x - cx)` where (cx, cy) is the centroid of the surface nodes. This gives a consistent circular ordering.

```python
def sort_by_arc_angle(surf_feats, surf_xy):
    """
    surf_feats: [B, N_surf, D]
    surf_xy: [B, N_surf, 2]  — surface node XY coordinates
    Returns: sorted feats [B, N_surf, D], inverse_idx [B, N_surf]
    """
    centroid = surf_xy.mean(dim=1, keepdim=True)  # [B, 1, 2]
    angles = torch.atan2(surf_xy[:,:,1] - centroid[:,:,1],
                         surf_xy[:,:,0] - centroid[:,:,0])  # [B, N_surf]
    sort_idx = angles.argsort(dim=1)  # [B, N_surf]
    inverse_idx = sort_idx.argsort(dim=1)  # to unsort later
    sorted_feats = torch.gather(surf_feats, 1,
                                sort_idx.unsqueeze(-1).expand_as(surf_feats))
    return sorted_feats, inverse_idx
```

### Step 2: Replace SRF MLP with bidirectional GRU

In `train.py`, find the `SurfaceRefinementHead` class (or equivalent). Replace the MLP layers with:

```python
class GRUSurfaceDecoder(nn.Module):
    def __init__(self, d_in, d_hidden=192, d_out=3, n_layers=2):
        super().__init__()
        self.gru = nn.GRU(
            input_size=d_in,
            hidden_size=d_hidden,
            num_layers=n_layers,
            batch_first=True,
            bidirectional=True,   # forward + backward pass
        )
        # bidirectional doubles output dim
        self.proj = nn.Linear(d_hidden * 2, d_out)

    def forward(self, x_sorted):
        """x_sorted: [B, N_surf, D] — sorted by arc-length"""
        out, _ = self.gru(x_sorted)   # [B, N_surf, 2*d_hidden]
        return self.proj(out)          # [B, N_surf, d_out]
```

Key parameters: `d_hidden=192` (matches current SRF hidden size), `n_layers=2`.

### Step 3: Integrate sort/unsort into forward pass

In the SRF forward pass:
```python
# Sort by arc-length
sorted_feats, inverse_idx = sort_by_arc_angle(surf_feats, surf_xy)

# GRU decode
pred_sorted = self.gru_decoder(sorted_feats)   # [B, N_surf, 3]

# Unsort back to original node order
pred = torch.gather(pred_sorted, 1,
                    inverse_idx.unsqueeze(-1).expand_as(pred_sorted))
```

Apply separately to fore-foil and aft-foil surface nodes (do NOT share GRU weights between foils — aft foil has different pressure dynamics).

### Step 4: Add flag

Add `--srf_gru` (bool flag). When set, replace the SRF MLP with GRU decoder. When not set, use original MLP (backward compat).

### Training commands (2 seeds)

```bash
CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent thorfinn --wandb_name "thorfinn/gru-decoder-s42" --wandb_group gru-sequential-surface-decoder \
  --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 \
  --srf_gru

CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent thorfinn --wandb_name "thorfinn/gru-decoder-s73" --wandb_group gru-sequential-surface-decoder \
  --seed 73 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 \
  --srf_gru
```

**Important**: baseline now includes Panel Cp flags (`--cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1`). Build on top of merged PR #2319.

### Implementation tips
- GRU is a standard PyTorch module — no new packages needed
- The bidirectional GRU has 2× parameters vs unidirectional but is still small (~1.5M params total for 2 layers, d_hidden=192)
- If surface_xy coordinates are not directly available at the SRF head, find them from the input batch — they should be in the first 2 features (x, y coordinates) of the node feature tensor
- Initialize GRU with default PyTorch initialization (orthogonal for recurrent weights) — this is stable

## Baseline

**Current best** (PR #2319, 2-seed avg):

| Metric | Value | Target |
|--------|-------|--------|
| p_in | 11.709 | < 11.709 |
| p_oodc | 7.544 | < 7.544 |
| p_tan | 27.402 | < 27.402 |
| p_re | 6.481 | < 6.481 |

W&B reference runs: h6fqcry4 (seed 42), cuhoscp9 (seed 73)

Reproduce baseline:
```bash
cd cfd_tandemfoil && python train.py \
  --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT thorfinn:

## Results — GRU Sequential Surface Decoder

Replaced the MLP-based Surface Refinement Head (SRF) with a bidirectional GRU that processes surface nodes in arc-length order (sorted by polar angle from centroid). Separate GRU heads for fore-foil and aft-foil surface nodes.

### Surface Pressure MAE (primary metric)

| Metric | Baseline (v3 Cp) | GRU Decoder | Δ% vs baseline |
|--------|-------------------|-------------|-----------------|
| **p_in** | 11.709 | **31.611** | **+170.0%** |
| **p_oodc** | 7.544 | **23.178** | **+207.2%** |
| **p_tan** | 27.402 | **41.961** | **+53.1%** |
| **p_re** | 6.481 | **19.203** | **+196.3%** |

### Per-seed breakdown

| Metric | s42 (z87dvheq) | s73 (5hsewpb3) | Avg |
|--------|----------------|----------------|-----|
| p_in | 34.756 | 28.466 | 31.611 |
| p_oodc | 25.633 | 20.722 | 23.178 |
| p_tan | 45.364 | 38.557 | 41.961 |
| p_re | 21.457 | 16.948 | 19.203 |
| Epochs | 39 | 40 | — |
| Time/epoch | 281s | 272s | ~277s |
| Peak VRAM | 41.1 GB | 41.1 GB | — |

### Experiment details
- **Command**: `python train.py --agent thorfinn --wandb_name "thorfinn/gru-decoder-s{42,73}" --wandb_group gru-sequential-surface-decoder --seed {42,73} --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --srf_gru`
- **W&B runs**: `z87dvheq` (seed 42), `5hsewpb3` (seed 73)
- **Epochs**: 39-40 (180-min wall-clock timeout)
- **Peak VRAM**: 41.1 GB

### Analysis — Clear negative result

**The GRU decoder severely degraded all metrics.** The primary cause is a **training speed bottleneck**:

- **Baseline**: ~68s/epoch → ~160 epochs in 180 min
- **GRU decoder**: ~277s/epoch → ~39 epochs in 180 min (4x slower)

The per-sample for-loop (sorting + GRU per foil per batch element) cannot be vectorized across the batch because surface node counts vary per sample. With only 39 epochs, the model never reached the convergence regime that baseline achieves around epoch 80-100.

**Secondary factors:**
1. **Default learning rate**: The GRU heads (2.2M new params) used the same LR as the backbone. Research suggests GRU heads need 5-10x higher LR to converge from random initialization when attached to a pre-trained backbone
2. **No orthogonal initialization**: Default PyTorch GRU init may destabilize long-sequence training. Orthogonal init of `weight_hh` matrices and reset gate bias = -1 are recommended best practices
3. **Cosine schedule mismatch**: `cosine_T_max=150` was tuned for 160 epochs — with only 39 epochs, the LR barely decays from peak, giving less stable late-epoch behavior

### Verdict

The hypothesis is **not falsified** but the implementation is **not viable in its current form** due to the per-sample loop bottleneck. The concept of sequential surface processing has theoretical merit (supported by Mamba Neural Operator results), but requires a fundamentally different implementation approach to be testable.

### Suggested follow-ups
1. **Batched GRU**: Pad surface node sequences to fixed length and process the entire batch in a single GRU call — this would eliminate the per-sample loop and restore ~68s/epoch throughput
2. **Higher GRU LR (5-10x)**: Separate param group with `lr = 2e-3` for GRU heads vs `2e-4` for backbone
3. **Orthogonal GRU init**: `nn.init.orthogonal_()` on `weight_hh` matrices, reset gate bias = -1
4. **1D convolution alternative**: Replace GRU with 1D causal convolutions on arc-length-sorted surface features — no recurrent state, fully parallelizable, and can still capture local sequential structure
5. **Attention-based surface decoder**: Small self-attention head on surface nodes only (much fewer tokens than full field) — retains global context without sequential bottleneck

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2338 because GRU decoder is 4x slower per epoch (277s vs 68s), only reaching 39 epochs vs 160 — never converges. All metrics +53-207% worse. Thorfinn's analysis is correct: the per-sample loop for variable-length surface nodes is the bottleneck. The sequential hypothesis has theoretical merit but needs a batched/padded implementation or a 1D conv alternative. Your suggestion #4 (1D causal conv) aligns with frieren's current experiment #2343. W&B: z87dvheq (s42), 5hsewpb3 (s73)..
